Allow to get existing ChannelOption / AttributeKey from String

normanmaurer opened the pull request:
Motivation:

We should allow to get a ChannelOption/AttributeKey from a String. This will make it a lot easier to make use of configuration files in applications.

Modifications:

- Add exists(...), createOrFail(...) method to ChannelOption and AttributeKey and alter valueOf(...) to return an existing instance for a String or create one.
- Add unit tests.

Result:

Much more flexible usage of ChannelOption and AttributeKey.

normanmaurer commented:
@trustin please review

Scottmitch commented on common/src/main/java/io/netty/util/AttributeKey.java:
```diff
@@ -29,14 +29,54 @@
+        if (name == null) {
+            throw new NullPointerException("name");
+        }
+        AttributeKey<T> option = names.get(name);
```
Consider reducing code:

```java
AttributeKey<T> option = names.get(checkNotNull(name, "name"));
```

Scottmitch commented on common/src/main/java/io/netty/util/AttributeKey.java:
```diff
@@ -29,14 +29,54 @@
+        if (name == null) {
+            throw new NullPointerException("name");
+        }
+        return names.containsKey(name);
```
Consider reducing code:

```java
return names.containsKey(checkNotNull(name, "name"));
```

Scottmitch commented on common/src/main/java/io/netty/util/AttributeKey.java:
```diff
@@ -29,14 +29,54 @@
+        if (name == null) {
+            throw new NullPointerException("name");
+        }
+        AttributeKey<T> option = new AttributeKey<T>(name);
```
`checkNotNull(...)`

Scottmitch commented on common/src/main/java/io/netty/util/UniqueName.java:
```diff
@@ -57,6 +57,15 @@ public UniqueName(ConcurrentMap<String, Boolean> map, String name, Object... arg
+            throw new NullPointerException("name");
+        }
+
+        id = nextId.incrementAndGet();
```
Reduce code a bit:

```java
this.name = checkNotNull(name, "name");
id = nextId.incrementAndGet();
```

Scottmitch commented on transport/src/main/java/io/netty/channel/ChannelOption.java:
```diff
@@ -85,10 +86,50 @@
+    @SuppressWarnings("unchecked")
     public static <T> ChannelOption<T> valueOf(String name) {
-        return new ChannelOption<T>(name);
+        if (name == null) {
```
Is there anyway to extract these new methods out and share this code with `AttributeKey`?  Perhaps a utility class with a factory member and an extra generic parameter would do the trick?

Scottmitch commented:
Done reviewing.

normanmaurer commented:
@Scottmitch addressed your comments accept code-sharing. The problem here is that we only lazy create the stuff which makes sharing a bit ugly. Let us just duplicate the code for now

Scottmitch commented:
@normanmaurer - Ok not ideal but it is only a few methods.  If it is too big of a PITA then I'm satisfied.

trustin commented on common/src/main/java/io/netty/util/AttributeKey.java:
```diff
@@ -29,14 +31,48 @@
+     * {@link IllegalArgumentException} if a {@link AttributeKey} for the given {@param name} exists.
+     */
+    @SuppressWarnings("unchecked")
+    public static <T> AttributeKey<T> createOrFail(String name) {
```
`newInstance`? We never used `create` in our code base.

trustin commented on transport/src/main/java/io/netty/channel/ChannelOption.java:
```diff
@@ -85,10 +88,44 @@
+     * {@link IllegalArgumentException} if a {@link ChannelOption} for the given {@param name} exists.
+     */
+    @SuppressWarnings("unchecked")
+    public static <T> ChannelOption<T> createOrFail(String name) {
```
Same naming issue.

normanmaurer replied:
you would like `newInstance` or `newInstanceOrFail` ?

trustin replied:
newInstance() - just like we don't name pipeline.addLast() as
addLastOrFail()

--
https://twitter.com/trustin https://twitter.com/trustin_ko
https://twitter.com/netty_project


On Wed, Feb 18, 2015, at 03:41 AM, Norman Maurer wrote:
> In transport/src/main/java/io/netty/channel/ChannelOption.java[1]:


> > +    }
> +
> +    /**
> +     * Returns {@code true} if a {@link ChannelOption} exists for the
>       given {@code name}.
> +     */
> +    public static boolean exists(String name) {
> +        checkNotNull(name, "name");
> +        return names.containsKey(name);
> +    } +
> +    /**
> +     * Creates a new {@link ChannelOption} for the given {@param
>       name} or fail with an
> +     * {@link IllegalArgumentException} if a {@link ChannelOption}
>       for the given {@param name} exists.
> +     */
> +    @SuppressWarnings("unchecked")
> +    public static <T> ChannelOption<T> createOrFail(String name) {
>      you would like newInstance or newInstanceOrFail ?


> — Reply to this email directly or view it on GitHub[2].

>



Links:

  1. https://github.com/netty/netty/pull/3421#discussion_r24840981
  2. https://github.com/netty/netty/pull/3421/files#r24840981

trustin commented:
Looks good except for the naming issue I mentioned.

normanmaurer commented:
Will address and cherry-pick



> Am 17.02.2015 um 12:02 schrieb Trustin Lee <notifications@github.com>:
> 
> Looks good except for the naming issue I mentioned.
> 
> —
> Reply to this email directly or view it on GitHub.
>

normanmaurer commented:
Cherry-picked into 4.0 (47f095856cebf1db6982d8a1e945ccca0ac32bb1), 4.1 (99bd43ed5163bf635ef4c9f1db2d93d7f0232382) and master (eba35bca0344b9b3e1497ccfa9ce0c638ca2ff49)
